### ISSUE TITLE
Index.name is not validatied to be Hashable

### ISSUE THREAD

**Issue by TomAugspurger** [pandas-dev/pandas#29069]
We require that Series.name be hashable. We don't do the same for Index.name. Should we?

```python
In [4]: pd.Index([], name=[])
Out[4]: Index([], dtype='object', name=[])
```

```python
In [5]: pd.Series([], name=[])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-2c73ddde103e> in <module>
----> 1 pd.Series([], name=[])

~/sandbox/pandas/pandas/core/series.py in __init__(self, data, index, dtype, name, copy, fastpath)
    326         generic.NDFrame.__init__(self, data, fastpath=True)
    327
--> 328         self.name = name
    329         self._set_axis(0, index, fastpath=True)
    330

~/sandbox/pandas/pandas/core/generic.py in __setattr__(self, name, value)
   5257             object.__setattr__(self, name, value)
   5258         elif name in self._metadata:
-> 5259             object.__setattr__(self, name, value)
   5260         else:
   5261             try:

~/sandbox/pandas/pandas/core/series.py in name(self, value)
    468     def name(self, value):
    469         if value is not None and not is_hashable(value):
--> 470             raise TypeError("Series.name must be a hashable type")
    471         object.__setattr__(self, "_name", value)
    472

TypeError: Series.name must be a hashable type
```

**Comment by WillAyd** [comment]
Makes sense - I would think generically any labels should have this validation

**Comment by R1j1t** [comment]
I was looking at this issue and would it be fine if i update [this line](https://github.com/pandas-dev/pandas/blob/master/pandas/core/indexes/base.py#L278) to 

Current version
```
        if name is None and hasattr(data, "name"):
            name = data.name

```

Proposed solution
```
        if name is None and hasattr(data, "name"):
            if not is_hashable(data.name):
                raise TypeError("Index.name must be a hashable type")
            name = data.name
```
I am happy to raise a PR after adding appropriate tests. 
> any labels should have this validation 

@WillAyd could you help me understand if you have any other function/labels in mind?

**Comment by TomAugspurger** [comment]
I'd recommend moving the Series.name implementation to `pandas.core.base.IndexOpsMixin`, which Series and Index inherit from.

**Comment by R1j1t** [comment]
Okay, so if i understood correctly i thought something like this should work 

```
    @property
    def name(self) -> Optional[Hashable]:
        return self.attrs.get("name", None)

    @name.setter
    def name(self, value: Optional[Hashable]) -> None:
        if not is_hashable(value):
            raise TypeError("name must be a hashable type")
        self.attrs["name"] = value

```

moving the above lines [from](https://github.com/pandas-dev/pandas/blob/master/pandas/core/series.py#L475) pandas.core.Series.py to  pandas.core.base.IndexOpsMixin should solve the problem but it only worked for series. I will try to look into your suggestion. Also my above thought didnt work.

**Comment by R1j1t** [comment]
I tried working with your suggestion of moving Series.name implementation to `pandas.core.base.IndexOpsMixin` but that would require a decent amount of changes (which I am not confident about). 

Also, I looked at the code of `pandas/core/indexes/base.py` and [here](https://github.com/pandas-dev/pandas/blob/v0.25.2/pandas/core/indexes/base.py#L1358-L1385) they validate `name` to be Hashable in `Index` class. Based on this I suggest the following:

Current version
```
        if name is None and hasattr(data, "name"):
            name = data.name
```
Proposed solution
```
        if name is None and hasattr(data, "name"):
            if not is_hashable(name):
                raise TypeError("Index.name must be a hashable type")
            name = data.name
        elif name is not None:
            if not is_hashable(name):
                raise TypeError("Index.name must be a hashable type")
```

If you still feel moving Series.name implementation to `pandas.core.base.IndexOpsMixin` would be better, then it is not straight forward to me and hence I am not proposing any solution for that. The problem I have, `Series` does the validation via `generic.NDFrame` and `NDFrame` is not inherited by `Index`.  So I will have to duplicate the checking directly to `IndexOpsMixin`, which I wanted to avoid. Let me know if I am wrong.
